Add decorations to ir nodes in Painless

jdconrad opened the pull request:
This PR starts to switch all ir nodes to use decorations instead of specific member data. For this specific PR, we add only an expression type decoration to begin with to keep the PR smaller. Other members of the ir nodes will be converted to decorations in a future PR.

The decoration system has two important advantages:
1) The first is it's consistent with the user nodes as generated data is attached to them as a decoration, so we have a clear, consistent model for how to use both trees.
2) This allows additionally generated data to be attached as necessary for optimization phases making the ir tree extendable which is one of our primary, long-term goals.

jdconrad committed:
make location final in IRNode

jdconrad committed:
add scoping for script, class, and method to WriteScope

jdconrad committed:
update scope to include classwriter and methodwriter

jdconrad committed:
move loop labels to writescope

jdconrad committed:
move try/catch labels to WriteScope

jdconrad committed:
move write to an external phase

jdconrad committed:
convert expression type to ir decoration

jdconrad committed:
Merge branch 'master' into proto

jdconrad committed:
Merge branch 'master' into proto

jdconrad committed:
Merge branch 'proto' into proto2

jdconrad committed:
Merge branch 'master' into proto

jdconrad committed:
Merge branch 'proto' into proto2

jdconrad committed:
response to pr comments

jdconrad committed:
Merge branch 'master' into proto2

jdconrad committed:
Merge branch 'proto2' into proto3

jdconrad committed:
Merge branch 'master' into proto2

jdconrad committed:
Merge branch 'proto2' into proto3

jdconrad committed:
response to pr comments

jdconrad committed:
Merge branch 'master' into proto3

jdconrad committed:
Merge branch 'master' into proto3

jdconrad committed:
Merge branch 'proto3' into proto4

elasticmachine commented:
Pinging @elastic/es-core-infra (:Core/Infra/Scripting)

stu-elastic commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultIRTreeToASMBytesPhase.java:
```diff
@@ -770,18 +772,22 @@ public void visitBinaryMath(BinaryMathNode irBinaryMathNode, WriteScope writeSco
+                        operation, irBinaryMathNode.getFlags());
             } else {
-                methodWriter.writeBinaryInstruction(irBinaryMathNode.getLocation(), irBinaryMathNode.getExpressionType(), operation);
+                methodWriter.writeBinaryInstruction(
```
expand this.

jdconrad replied:
Done.

stu-elastic commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultIRTreeToASMBytesPhase.java:
```diff
@@ -990,15 +997,16 @@ public void visitInstanceof(InstanceofNode irInstanceofNode, WriteScope writeSco
             Class<?> boxedInstanceType = PainlessLookupUtility.typeToBoxedType(instanceType);
-            Class<?> boxedExpressionType = PainlessLookupUtility.typeToBoxedType(irChildNode.getExpressionType());
+            Class<?> boxedExpressionType =
+                    PainlessLookupUtility.typeToBoxedType(irChildNode.getDecoration(IRDExpressionType.class).getType());
```
pull into intermediate variable.

jdconrad replied:
Done.

stu-elastic commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultIRTreeToASMBytesPhase.java:
```diff
@@ -1195,11 +1203,11 @@ public void visitTypeCaptureReference(TypedCaptureReferenceNode irTypedCaptureRe
-        Type methodType = Type.getMethodType(
-                MethodWriter.getType(irTypedCaptureReferenceNode.getExpressionType()),
+        Type methodType = Type.getMethodType(MethodWriter.getType(
+                irTypedCaptureReferenceNode.getDecoration(IRDExpressionType.class).getType()),
```
pull into intermediate variable.

jdconrad replied:
Done.

stu-elastic commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultIRTreeToASMBytesPhase.java:
```diff
@@ -1195,11 +1203,11 @@ public void visitTypeCaptureReference(TypedCaptureReferenceNode irTypedCaptureRe
-        methodWriter.invokeDefCall(irTypedCaptureReferenceNode.getMethodName(),
-                methodType, DefBootstrap.REFERENCE, irTypedCaptureReferenceNode.getExpressionCanonicalTypeName());
+        methodWriter.invokeDefCall(irTypedCaptureReferenceNode.getMethodName(), methodType, DefBootstrap.REFERENCE,
+                irTypedCaptureReferenceNode.getDecoration(IRDExpressionType.class).getCanonicalTypeName());
```
pull into intermediate variable.

jdconrad replied:
Done.

stu-elastic commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:
```diff
@@ -973,14 +975,15 @@ public void visitBinary(EBinary userBinaryNode, ScriptScope scriptScope) {
         BooleanNode irBooleanNode = new BooleanNode(userBooleanCompNode.getLocation());
-        irBooleanNode.setExpressionType(scriptScope.getDecoration(userBooleanCompNode, ValueType.class).getValueType());
+        irBooleanNode.attachDecoration(
+                new IRDExpressionType(scriptScope.getDecoration(userBooleanCompNode, ValueType.class).getValueType()));
```
pull into intermediate variable.

jdconrad replied:
Done.

stu-elastic commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:
```diff
@@ -1008,7 +1011,8 @@ public void visitExplicit(EExplicit userExplicitNode, ScriptScope scriptScope) {
         InstanceofNode irInstanceofNode = new InstanceofNode(userInstanceofNode.getLocation());
-        irInstanceofNode.setExpressionType(scriptScope.getDecoration(userInstanceofNode, ValueType.class).getValueType());
+        irInstanceofNode.attachDecoration(
+                new IRDExpressionType(scriptScope.getDecoration(userInstanceofNode, ValueType.class).getValueType()));
```
pull into intermediate variable.

jdconrad replied:
Done.

stu-elastic commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:
```diff
@@ -1091,7 +1098,8 @@ public void visitNewArray(ENewArray userNewArrayNode, ScriptScope scriptScope) {
 
-        irNewObjectNode.setExpressionType(scriptScope.getDecoration(userNewObjectNode, ValueType.class).getValueType());
+        irNewObjectNode.attachDecoration(
+                new IRDExpressionType(scriptScope.getDecoration(userNewObjectNode, ValueType.class).getValueType()));
```
pull into intermediate variable.

jdconrad replied:
Done.

stu-elastic commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:
```diff
@@ -1151,15 +1159,17 @@ public void visitCallLocal(ECallLocal callLocalNode, ScriptScope scriptScope) {
 
-        irInvokeCallMemberNode.setExpressionType(scriptScope.getDecoration(callLocalNode, ValueType.class).getValueType());
+        irInvokeCallMemberNode.attachDecoration(
+                new IRDExpressionType(scriptScope.getDecoration(callLocalNode, ValueType.class).getValueType()));
```
pull into intermediate variable.

jdconrad replied:
Done.

stu-elastic commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:
```diff
@@ -1151,15 +1159,17 @@ public void visitCallLocal(ECallLocal callLocalNode, ScriptScope scriptScope) {
         ConstantNode irConstantNode = new ConstantNode(userBooleanConstantNode.getLocation());
-        irConstantNode.setExpressionType(scriptScope.getDecoration(userBooleanConstantNode, ValueType.class).getValueType());
+        irConstantNode.attachDecoration(
+                new IRDExpressionType(scriptScope.getDecoration(userBooleanConstantNode, ValueType.class).getValueType()));
```
pull into intermediate variable.

jdconrad replied:
Done.

stu-elastic commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:
```diff
@@ -1168,7 +1178,8 @@ public void visitBooleanConstant(EBooleanConstant userBooleanConstantNode, Scrip
         ConstantNode irConstantNode = new ConstantNode(userNumericNode.getLocation());
-        irConstantNode.setExpressionType(scriptScope.getDecoration(userNumericNode, ValueType.class).getValueType());
+        irConstantNode.attachDecoration(
+                new IRDExpressionType(scriptScope.getDecoration(userNumericNode, ValueType.class).getValueType()));
```
pull into intermediate variable.

jdconrad replied:
Done.

stu-elastic commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:
```diff
@@ -1177,7 +1188,8 @@ public void visitNumeric(ENumeric userNumericNode, ScriptScope scriptScope) {
         ConstantNode irConstantNode = new ConstantNode(userDecimalNode.getLocation());
-        irConstantNode.setExpressionType(scriptScope.getDecoration(userDecimalNode, ValueType.class).getValueType());
+        irConstantNode.attachDecoration(
+                new IRDExpressionType(scriptScope.getDecoration(userDecimalNode, ValueType.class).getValueType()));
```
pull into intermediate variable.

jdconrad replied:
Done.

stu-elastic commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:
```diff
@@ -1335,7 +1347,8 @@ public void visitFunctionRef(EFunctionRef userFunctionRefNode, ScriptScope scrip
 
-        irReferenceNode.setExpressionType(scriptScope.getDecoration(userFunctionRefNode, ValueType.class).getValueType());
+        irReferenceNode.attachDecoration(
+                new IRDExpressionType(scriptScope.getDecoration(userFunctionRefNode, ValueType.class).getValueType()));
```
pull into intermediate variable.

jdconrad replied:
Done.

stu-elastic commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:
```diff
@@ -1393,7 +1406,8 @@ public void visitNewArrayFunctionRef(ENewArrayFunctionRef userNewArrayFunctionRe
 
-        irReferenceNode.setExpressionType(scriptScope.getDecoration(userNewArrayFunctionRefNode, ValueType.class).getValueType());
+        irReferenceNode.attachDecoration(
+                new IRDExpressionType(scriptScope.getDecoration(userNewArrayFunctionRefNode, ValueType.class).getValueType()));
```
pull into intermediate variable.

jdconrad replied:
Done.

stu-elastic commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:
```diff
@@ -1665,20 +1679,21 @@ public void visitBrace(EBrace userBraceNode, ScriptScope scriptScope) {
                 LoadMapShortcutNode irLoadMapShortcutNode = new LoadMapShortcutNode(location);
-                irLoadMapShortcutNode.setExpressionType(scriptScope.getDecoration(userBraceNode, ValueType.class).getValueType());
+                irLoadMapShortcutNode.attachDecoration(
+                        new IRDExpressionType(scriptScope.getDecoration(userBraceNode, ValueType.class).getValueType()));
```
pull into intermediate variable.

jdconrad replied:
Done.

stu-elastic commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:
```diff
@@ -1687,7 +1702,8 @@ public void visitBrace(EBrace userBraceNode, ScriptScope scriptScope) {
                 LoadListShortcutNode irLoadListShortcutNode = new LoadListShortcutNode(location);
-                irLoadListShortcutNode.setExpressionType(scriptScope.getDecoration(userBraceNode, ValueType.class).getValueType());
+                irLoadListShortcutNode.attachDecoration(
+                        new IRDExpressionType(scriptScope.getDecoration(userBraceNode, ValueType.class).getValueType()));
```
pull into intermediate variable.

jdconrad replied:
Done.

stu-elastic commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:
```diff
@@ -1715,7 +1731,8 @@ public void visitCall(ECall userCallNode, ScriptScope scriptScope) {
 
-            irCallSubDefNode.setExpressionType(scriptScope.getDecoration(userCallNode, ValueType.class).getValueType());
+            irCallSubDefNode.attachDecoration(
+                    new IRDExpressionType(scriptScope.getDecoration(userCallNode, ValueType.class).getValueType()));
```
pull into intermediate variable.

jdconrad replied:
Done.

stu-elastic commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:
```diff
@@ -1751,7 +1768,8 @@ public void visitCall(ECall userCallNode, ScriptScope scriptScope) {
 
-            irInvokeCallNode.setExpressionType(scriptScope.getDecoration(userCallNode, ValueType.class).getValueType());;
+            irInvokeCallNode.attachDecoration(
+                    new IRDExpressionType(scriptScope.getDecoration(userCallNode, ValueType.class).getValueType()));
```
pull into intermediate variable.

jdconrad replied:
Done.

stu-elastic reviewed (approved):
Create an issue to be able to print out the tree, with decorations, in every phase.

jdconrad committed:
Merge branch 'master' into proto3

jdconrad committed:
Merge branch 'proto3' into proto4

jdconrad committed:
response to pr comments

jdconrad commented:
@stu-elastic Thanks for the review! Updated each place you mentioned in the PR to make more local vars for readability. Added (https://github.com/elastic/elasticsearch/issues/64076) for printing trees. Will commit as soon as CI passes.
